source/Calamari/TargetDiscoveryBehaviour.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Calamari.Azure;
+////using Calamari.AzureAppService.Azure;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Discovery;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Pipeline;
+using Calamari.Common.Plumbing.ServiceMessages;
+using Calamari.Common.Plumbing.Variables;
+using Microsoft.Azure.Management.AppService.Fluent;
+
+#nullable enable
+namespace Calamari.AzureAppService
+{
+    public class TargetDiscoveryBehaviour : IDeployBehaviour
+    {
+        private ILog Log { get; }
+
+        public TargetDiscoveryBehaviour(ILog log)
+        {
+            Log = log;
+        }
+
+        public bool IsEnabled(RunningDeployment context) => true;
+
+        public async Task Execute(RunningDeployment runningDeployment)
+        {
+            await Task.CompletedTask;
+            var targetDiscoveryContext = GetTargetDiscoveryContext(runningDeployment.Variables);
+            if (targetDiscoveryContext == null)
+            {
+                Log.Warn("Aborting target discovery.");
+                return;
+            }
+
+            var account = targetDiscoveryContext.Authentication.AccountDetails;
+            var azureClient = account.CreateAzureClient();
+
+            try
+            {
+                var webApps = azureClient.WebApps.ListWebAppBasic();
+                foreach (var webApp in webApps)
+                {
+                    var tags = GetTags(webApp);
+                    var matchResult = targetDiscoveryContext.Scope.Match(tags);
+                    if (matchResult.IsSuccess)
+                    {
+                        Log.Info($"Discovered matching web app: {webApp.Name}");
+                        WriteTargetCreationServiceMessage(
+                            webApp, targetDiscoveryContext, matchResult, runningDeployment.Variables);
+                    }
+                }
+
+                if (!webApps.Any())
+                {
+                    Log.Info("No matching web apps found.");
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Warn($"Error connecting to Azure to look for web apps:");
```

Comment:
I'm trying to through cases of where we might want a failure during discovery to bubble out to the user and therefore fail the deployment or potentially kick into the guided failure handling. 

I wonder what the user would expect to have happen if say we didn't find any new targets to deploy to because of a discovery error, and there were some existing targets that got deployed to but the new ones they expected to didn't? There would be warnings in the deployment but do users actually take too much notice of those?

Reply:
We _could_ possibly try to distinguish configuration errors from potentially transient connectivity errors.
E.g. auth failure versus connectivity failure.
I.e. For auth failures due to bad account credentials indicates a configuration error and we halt the deployment until the user has fixed the configuration, but for transient connectivity failures we push on with the deployment.

I'm still leaning towards not halting the deployment due to target discovery failures even if they are down to configuration errors. E.g. consider a case where a target has been previously discovered, but the the service principal used for discovery is removed from Azure. Do we really want to halt further deployments to the existing target until the account used for discovery is fixed? (Assuming account used for deployment is different or fixed etc.)

What is your opinion @rhysparry ?

Reply:
I agree that we shouldn't halt the deployment process (although if the step itself resolves no targets, that should probably fail). It would be good if we could configure this per project in the future.

Reply:
> although if the step itself resolves no targets, that should probably fail
I don't believe it works this way at the moment, no targets results in the step being skipped.

> It would be good if we could configure this per project in the future.
Guided failure mode gives some level of configuration I suppose, the user can choose to continue or fail on error.

Reply:
Here's a situation: what if they have a manually created a target for a specific environment but have dynamic deployment enabled (globally) for another environment. They go to deploy there configuration but discovery returns an empty array so the deployment fails, even though their target exists and is healthy, it's just not tagged so it doesn't show up in discovery. Do we force the user to tag those assets so that discovery works?

I suggest the final step should fail if there are no targets to deploy to when dynamic discovery is switched on. (maybe it could be configurable, but on by default). Maybe this wouldn't help if you had targets already but expected more to come in...

Reply:
> Here's a situation: what if they have a manually created a target for a specific environment but have dynamic deployment enabled (globally) for another environment. They go to deploy there configuration but discovery returns an empty array so the deployment fails, even though their target exists and is healthy, it's just not tagged so it doesn't show up in discovery. Do we force the user to tag those assets so that discovery works?

Do you mean they are deploying to the environment that doesn't have discovery enabled? If so, discovery wouldn't happen for deployments to that environment. Does that help to cater for this situation?

Reply:
> I suggest the final step should fail if there are no targets to deploy to when dynamic discovery is switched on. (maybe it could be configurable, but on by default). Maybe this wouldn't help if you had targets already but expected more to come in...

I agree. Our change to the defaults to allow deployments to be created without targets might have given some safeguards before. If you were doing script discovery you could choose whether that step fails or not if no targets are discovered. No targets found _and_ no targets matching feels like we need to escalate.

Is it feasible to fail only in this scenario? (We did a discovery, found nothing (or it failed) and then we had no matching targets pre-existing?)

This should probably be user-configurable, but happy to push that off unless that makes us more comfortable taking a position on this.

Reply:
> Do you mean they are deploying to the environment that doesn't have discovery enabled? If so, discovery wouldn't happen for deployments to that environment. Does that help to cater for this situation?

Ah yes sorry, I forgot that environments can turn off discovery... I guess it depends, do we think the situation could happen where no targets are returned from discovery but there is a valid target ready and waiting thus failing on the discovery step would jump the gun and prevent a normal successful deployment?

>Is it feasible to fail only in this scenario? (We did a discovery, found nothing (or it failed) and then we had no matching targets pre-existing?)

I feel like this is feasible... (I could be wrong). I think we would just need to know if discovery is enabled in the final deployment step and if it is, and we have no valid/healthy targets, the step fails. How we make this configurable is a different matter...

Reply:
This is a wider question than just error handling during target discovery.
I've created a new Trello card to cover it: https://trello.com/c/9GSW7KAc/34-deployments-that-that-end-up-with-no-targets-should-be-treatable-as-failures
I don't expect we can address it in this PR.

Reply:
I have also added a warning if no targets are found during discovery, as I think if target discovery is configured then we would expect it to find targets almost all the time.
It still won't halt the deployment or treat it as a failure, but at least there will be some clue for the user instead of fully-green deployment output.